src/snowflake/snowpark/mock/_functions.py

```diff
@@ -929,3 +930,101 @@ def mock_to_variant(expr: ColumnEmulator):
     res = expr.copy()
     res.sf_type = ColumnType(VariantType(), expr.sf_type.nullable)
     return res
+
+
+# More info about all allowed aliases here:
+# https://docs.snowflake.com/en/sql-reference/functions-date-time#label-supported-date-time-parts
+
+DATETIME_PART_TO_ALIASES = {
+    "year": {"year", "y", "yy", "yyy", "yyyy", "yr", "years", "yrs"},
+    "quarter": {"quarter", "q", "qtr", "qtrs", "quarters"},
+    "month": {"month", "mm", "mon", "mons", "months"},
+    "week": {"week", "w", "wk", "weekofyear", "woy", "wy"},
+    "day": {"day", "d", "dd", "days", "dayofmonth"},
+    "hour": {"hour", "h", "hh", "hr", "hours", "hrs"},
+    "minute": {"minute", "m", "mi", "min", "minutes", "mins"},
+    "second": {"second", "s", "sec", "seconds", "secs"},
+    "millisecond": {"millisecond", "ms", "msec", "milliseconds"},
+    "microsecond": {"microsecond", "us", "usec", "microseconds"},
+    "nanosecond": {
+        "nanosecond",
+        "ns",
+        "nsec",
+        "nanosec",
+        "nsecond",
+        "nanoseconds",
+        "nanosecs",
+        "nseconds",
+    },
+}
+
+DATETIME_PARTS = set(DATETIME_PART_TO_ALIASES.keys())
+ALIASES_TO_DATETIME_PART = {
+    v: k for k, l in DATETIME_PART_TO_ALIASES.items() for v in l
+}
+DATETIME_ALIASES = set(ALIASES_TO_DATETIME_PART.keys())
+
+
+def unalias_datetime_part(part):
+    if part in DATETIME_ALIASES:
+        return ALIASES_TO_DATETIME_PART[part]
+    else:
+        raise ValueError(f"{part} is not a recognized date or time part.")
+
+
+def cast_to_datetime(date):
+    return datetime.datetime.fromordinal(date.toordinal())
+
+
+def add_years(date, duration):
+    return date.replace(year=date.year + duration)
+
+
+def add_months(scalar, date, duration):
+    # Keep months in 1..12
+    month = (date.month - 1 + scalar * duration) % 12 + 1
+
+    # Add years as needed
+    year = date.year + (date.month - 1 + scalar * duration) // 12
+
+    return date.replace(year=year, month=month)
+
+
+def add_timedelta(unit, date, duration, scalar=1):
+    return date + datetime.timedelta(**{f"{unit}s": duration * scalar})
+
+
+@patch("dateadd")
+def mock_dateadd(part: str, expr1: ColumnEmulator, expr2: ColumnEmulator):
+    # Extract a standardized name
+    part = unalias_datetime_part(part)
+    sf_type = expr1.sf_type
+    ts_type = ColumnType(TimestampType(TimestampTimeZone.NTZ), expr1.sf_type.nullable)
+
+    def nop(x):
+        return x
+
+    cast = nop
+
+    # Create a lambda that applies the transformation
+    # If the time unit is smaller than a day date types will be cast to datetime types
+    if part == "year":
+        func = add_years
+    elif part == "quarter" or part == "month":
+        scalar = 3 if part == "quarter" else 1
+        func = partial(add_months, scalar)
```

Comment:
I've structured this function into a section that determines the transform needed and a part that applies the transform. I like this pattern because it makes it clear that all cases have a transform applied in a similar way. Originally I had the transform function defined as a lambda one liner in each case, but black did not like that when styling the code.

Integrating the logic into each clause would require removing the separation of transform selection and application or defining functions inside of an if statement which I don't particularly like. For organization I could move all the helpers to the same level as `nop`, but that would mean recreating those functions every time `mock_dateadd` is called. I'm not opposed to that as I doubt that's a significant performance hit.

Reply:
I see, the pattern makes sense to me now. Shall we add a comment or two to explain the structure to help understanding?